### PR TITLE
Add method to update position of arrow patch

### LINKED ISSUE #14217: [Feature request] Add a way to update the position of the Arrow patch.
Hi,

Currently, the `arrow` class doesn't let you update the coordinates after creating the patch. That is, one **must** remove the old patch and create a new one just to move the arrows. This seems fairly inefficient.

Would it be possible to add some setter methods to let one change the arrow position and size without recreating a new one?

The motivation for this is that removing and adding a new patch requires a full re-draw (afaik) which is slow.

Thanks!

**Matplotlib version**

  * Operating system: Windows 10
  * Matplotlib version: 3.0.3
  * Matplotlib backend (`print(matplotlib.get_backend())`): qt5agg
  * Python version: 3.7

### PR DESCRIPTION
**Pull request by saranti**
## PR summary

Closes #14217. Adds `set_data` method from `FancyArrow` that allows the arrow to be repositioned without re-drawing.
Arrow's `x,` `y,` `dx`, `dy` and `width` parameters are now stored in private variables so that they can be reused for recreating the transform.

e.g.
```py
import matplotlib as mpl
import matplotlib.pyplot as plt
from matplotlib.patches import Arrow

fig, ax = plt.subplots()
ax.plot([0, 4], [10, 0])

a = mpl.patches.Arrow(2, 0, 0, 10)
ax.add_patch(a)

# code for modifing the arrow
a.set_data(x=.5, y=0, dx=3, dy=8, width=1.2)

plt.show()
```

See the discussion here https://github.com/matplotlib/matplotlib/issues/14217#issuecomment-538072364. I don't know what the consensus is on the Arrow class but If this fix is still needed, I'll add a test and a what's new entry.
## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Comment by timhoffm** [comment]
While the arrow patch has fundamental flaws, it's not going to go away soon. So this addition is reasonable.

File: lib/matplotlib/patches.py [hunk 1]
```diff
@@ -1297,11 +1297,16 @@ def __init__(self, x, y, dx, dy, *, width=1.0, **kwargs):
...
+        self._width = width
         self._patch_transform = (
             transforms.Affine2D()
-            .scale(np.hypot(dx, dy), width)
-            .rotate(np.arctan2(dy, dx))
-            .translate(x, y)
+            .scale(np.hypot(self._dx, self._dy), self._width)
+            .rotate(np.arctan2(self._dy, self._dx))
+            .translate(self._x, self._y)
             .frozen())
```
[the remark this answers is not in the document]
Comment (saranti) [reply lib/matplotlib/patches.py]: Yes, good idea

File: doc/users/next_whats_new/update_arrow_patch.rst [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    from matplotlib.patches import Arrow
+
+    fig, ax = plt.subplots()
+    ax.plot([0, 4], [10, 0])
+
+    a = mpl.patches.Arrow(2, 0, 0, 10)
+    ax.add_patch(a)
+
+    # code for modifying the arrow
+    a.set_data(x=.5, dx=3, dy=8, width=1.2)
```
[the remark this answers is not in the document]
Comment (saranti) [reply doc/users/next_whats_new/update_arrow_patch.rst:21]: That sounds good. I'll have a go at making an animation

File: lib/matplotlib/patches.py [hunk 3]
```diff
@@ -1297,19 +1297,47 @@ def __init__(self, x, y, dx, dy, *, width=1.0, **kwargs):
...
+        Set `.Arrow` x, y, dx, dy and width.
+        Values left as None will not be updated.
+
+        Parameters
+        ----------
+        x, y : float or None, default: None
+            The x and y coordinates of the arrow base.
+
+        dx, dy : float or None, default: None
+            The length of the arrow along x and y direction.
```
[the remark this answers is not in the document]
Comment (timhoffm) [reply lib/matplotlib/patches.py:1319]: That's not necessary. [numpydoc explicitly allows this](https://numpydoc.readthedocs.io/en/latest/format.html#parameters):

> When two or more input parameters have exactly the same type, shape and description, they can be combined:

**Review by timhoffm — APPROVED** [review]

**Review by dstansby — APPROVED** [review]

**Pull request merged**